### ISSUE TITLE
Webpack Failed to compile (create-react-app)

### ISSUE THREAD

**Issue by ksloan** [Automattic/mongoose#6826]
Redacted -- see https://mongoosejs.com/docs/browser.html

**Comment by ksloan** [comment]
I read http://mongoosejs.com/docs/browser.html -- and I understand how browser.js is being loaded using a webpack config.

However, is there a way to build mongoose on it's own prior to import so I can include it in a create-react-app build (since I can't access the webpack config without ejecting).

**Comment by ksloan** [comment]
SOLVED: compiled browser version is here https://github.com/zensurance/mongoose-browser/blob/master/index.js if anyone wants it

**Comment by vkarpov15** [comment]
@ksloan I wouldn't recommend using Mongoose 4.x on the client but Mongoose 5.x on the server if that's what you're planning on doing.

**Comment by arnaudriegert** [comment]
I'm not sure why this issue is closed (and redacted): the link you posted in your last edit doesn't quite answer the issue. On the contrary, it gives the impression that a partial version Mongoose can be used in the browser, which is true, but using this version doesn't solve the bug you encountered and that till seems valid to me.

I imagine I am not the only one in this case: I like the idea of having Mongoose schemas accessible in my React app so I can pre-validate the models on the client side before making a request to the server.
But when I try to bundle it using `create-react-app`'s `npm run build`, I have the error you originally posted: Mongoose (or at least, `lib/utils.js`) couldn't be minified. It works fine if I downgrade to v4.13.14 but not on v5. It also works fine in the development build (unminified) using just `npm start`.

I'm not an expert in Webpack, Babel and the differences between ES5 and ES6 so I'm not sure if this applies, but in [this comment](https://github.com/facebook/create-react-app/issues/1125#issuecomment-264195015) @gaearon indicates that it is the library authors who should take care of the compilation to make sure that the libraries work in the environments they want to support. This seemed like a valid point to me.

Would it make sense to make it so that the v5's `mongoose/browser` could be minified in `create-react-app`'s build step?

**Comment by vkarpov15** [comment]
@arnaudriegert OPs comment points to the legacy browser docs, and OP was the one who redacted it, so that was the reason for my response. Can you post what error you're seeing?

**Comment by ksloan** [comment]
Sorry I linked to wrong docs -- here is the updated docs that basically say "compile it yourself" https://mongoosejs.com/docs/browser.html

**Comment by vkarpov15** [comment]
Looking more closely at this, it looks like I can compile Mongoose 5.1.8, 5.2.18, and 5.3.3 with `npm run build` and `npm start` with just adding `import mongoose from 'mongoose/browser';` to the output of create-react-app v1.1.0.

![image](https://user-images.githubusercontent.com/1620265/46909373-2bb40300-ceff-11e8-9975-41f111e37829.png)

Can you please clarify what version of Mongoose you're using and double check that you're getting this error from just the normal output of `create-react-app`, without any changes to the webpack config?

**Comment by saiichihashimoto** [comment]
Using create-react-app, mongoose 5.3.4 builds in the browser whether importing mongoose or mongoose/browser.

Also, using create-react-app, mongoose 5.3.5 does NOT build in the browser whether importing mongoose or mongoose/browser. The error seems related to NodeBuffer, possibly because of #7102.

Unless ejecting from create-react-app, you don't have access to change the webpack & babel configs, so changing the webpack settings for buffer isn't really a solution.

**Comment by lineus** [comment]
@saiichihashimoto you can follow #7173 for the browser buffer issue. This issue is related to older versions of mongoose. The problem you are describing was introduced in 5.3.5.

@vkarpov15 
With regard to the issue with react/Buffer pointed out by @saiichihashimoto , It's worth noting that `npm run build` passes without error on 5.3.{5,6}, but `npm run start` shows the problem in the browser: `TypeError: NodeBuffer is undefined; can't access its "prototype" property`.

**Comment by vkarpov15** [comment]
That's a good point, we'll look into this.

**Comment by vkarpov15** [comment]
@saiichihashimoto that's a good point. Moving to telling users to compile it themselves was not a great idea in hindsight. We really only wanted to move away from hosting a compiled version on cloudfront. Something to add for the future

**Comment by saiichihashimoto** [comment]
I find that, while there's a lot of code reuse, you need to consciously think about how you want users to consume your library in the browser. I think this is mostly there! Do you know how mongoose was compiled for the browser in the past? webpack? babel? I might put in a PR to have deployments include a compiled version and have tests validate that the browser build is valid.

**Comment by vkarpov15** [comment]
@saiichihashimoto we used plain old browserify: https://github.com/Automattic/mongoose/blob/4.x/Makefile#L64-L67 . Keep in mind this code was from 2014, so it predates the widespread adoption of babel and webpack.

My only concern is that people may have different concerns re: browser compatibility. Mongoose will never work on ES4, and if I recall correctly there's some issue that prevents us from supporting IE8. Running tests on sauce was pretty flakey in the past, so I'm wary of running real browser tests. Are there better options out there now?

**Comment by saiichihashimoto** [comment]
Oh, I wasn't thinking of running browser tests. We could use whatever testing framework purely to see if we could make a schema, make a document, and run validateSync using the build. So webpack/babel is mostly likely the best options here.

**Comment by vkarpov15** [comment]
Yeah exactly. I suppose we can just rely on babel to do the right thing :)
